### ISSUE TITLE
Roslyn doesn't process [MemberNotNull] annotations from deconstruction calls in pattern matching (it does for other members e.g. properties)

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51934]
**Version Used**: 
```
Branch main (12 Mar 2021)
Latest commit 8da536d by msftbot[bot]:
Merge pull request #51811 from CyrusNajmabadi/simplifyCompletion

Simplify portions of symbol completion
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

public class Test
{
  [MemberNotNull(nameof(Value))] public void Deconstruct(out int x, out int y) => x = y = (Value ??= "").Length;
  [MemberNotNull(nameof(Value))] public int Prop => (Value ??= "").Length;
  
  public string? Value { get; set; }

  public void M1()
  {
    switch (this)
    {
      case (1, 1):
        Value.ToString(); // CS8602: Dereference of a possibly null reference.
        break;
    }
  }
    
  public void M2()
  {
    switch (this)
    {
      case { Prop: 1 }:
        Value.ToString(); // no warnings
        break;
    }
  }
}
```

**Expected Behavior**:
No warnings in both methods. Both methods invoke members with `[MemberNotNull(nameof(Value))]` annotations before using the `Value` property

**Actual Behavior**:
False `CS8602: Dereference of a possibly null reference.` is reported for `M1` but not `M2`

**Notes**
`NullableWalker.LearnFromDecisionDag` calls `ApplyMemberPostConditions(inputSlot, property.GetMethod);` for `BoundDagPropertyEvaluation` but not for `BoundDagDeconstructEvaluation` so it correctly processes `[MemberNotNull(When)]` annotations from properties but completely ignores such annotations on deconstruction methods

**Comment by RikkiGibson** [comment]
This also appears to be the case for a simple deconstruction assignment: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQtAAgAwn0CMAdACJQQDmMAewDOAFygBjIewDCAhEQCCFHAE8hUIQG5aDAMwsATCWkkA3rRKWW+towD8JAAoAnAQAczJXkRGaSQnz8AXwsrUMsAbQBZIjBgImcAOQERRNwcAAoKMCIBADMMl3cASmKAXXDrFgAWEk4icQEYUWcscREMgSwRElgegA84Ei6evpIVYpIAXgA+EgzB8aGit0mp+dYhgyGAIggd4u0aKyr6WqiMyfNjk6sANwhnecWJ6ZIRAAsNI9urFfYACoCADKImcsF4lx+VhCNFhQA===)

```cs
#nullable enable

using System.Diagnostics.CodeAnalysis;

public class C {
    public string? Prop { get; set; }
    
    [MemberNotNull(nameof(Prop))]
    public void Deconstruct(out int x, out int y) => (x, y, Prop) = (1, 2, "a");
    public void M() {
        var (x, y) = this;
        Prop.ToString(); // unexpected warning
    }
}

```
